### PR TITLE
[None] [feat] Enable run_post_quant_allgather for MoE TRTLLM backend

### PR DESCRIPTION
**Pull request by ChristinaZ**
### Add topk_ids and topk_weight as inputs for moe trtllgen backend
Add the two inputs `topk_ids` and `topk_weights` first. 
Then, enable post-quantization allgather for the Moe TrtllmGen backend.

1. Routing kernels
I modified the kernels for DeepSeek and Llama 4 and renormalized routing methods in files `RoutingDeepSeek.cu`, `RoutingRenormalize.cu` and `RoutingLlama4.cu`

- Related C++ unit tests are also added in files like `routingTest.cpp`.
- Besides that, I updated the default value of `maxExpertIdx` in file `RoutingLlama4.cu`, so that it can handle `NaN` correctly.
- Added more detailed type checks like this:
```
        if (static_cast<RoutingMethodType>(routing_method_type) == RoutingMethodType::DeepSeekV3)
        {
            TORCH_CHECK(routing_logits.value().scalar_type() == at::ScalarType::Float, "routing_logits must be float");
        }
        else
        {
            TORCH_CHECK(
                routing_logits.value().scalar_type() == at::ScalarType::BFloat16, "routing_logits must be bfloat16");
        }
```
- Changed the names of several variables to add new inputs (for example, changed `mPtrExpertIdx` to `mPtrTopKPacked`).

Hi @MatthiasKohl and @nekorobov, could you please review this part?

2. Add two parameters (`topk_ids` and `topk_weights`) in all the MoE trtllm-gen Ops. For example:
https://gitlab-master.nvidia.com/ftp/tekit/-/merge_requests/9611/diffs?file=1e58f3c66b27c6a2f10bc87b0a3381f3d846762c#1e58f3c66b27c6a2f10bc87b0a3381f3d846762c_673_731

Related unit tests in file `tests/unittest/_torch/thop/test_moe.py` 

3. @DomBrown  Hi Dom, this is the same PR you helped review before.

4.  Hi Daniel, as we are going to reuse the routing part for the post-quant all-gather of MoE Trtllm Gen backend, I might need BF16 output for Trtllm Gen backend. I added a parameter in the routing method for CUTLASS. Maybe you can take a look?  (For example [routing.py](https://gitlab-master.nvidia.com/ftp/tekit/-/blob/user/christinaz/add_inputs_moe_trtllmgen/tensorrt_llm/_torch/modules/fused_moe/routing.py?ref_type=heads#L222))

5. @rosenrodt  Hi Anthony, I added the post-quant allgather logic. Please help review it [fused_moe_trtllm_gen.py](https://gitlab-master.nvidia.com/ftp/tekit/-/blob/user/christinaz/add_inputs_moe_trtllmgen/tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py?ref_type=heads#L210)

## Test Coverage

```
cd cpp/build
make -j$(nproc) google-tests
./tests/unit_tests/kernels/routingKernelsTest

pytest -v -s tests/unittest/_torch/thop/test_moe.py

pytest -s -o log_cli=true "tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_nvfp4_4gpus[moe_backend=TRTLLM-mtp_nextn=2-tp4-fp8kv=True-attention_dp=True-cuda_graph=True-overlap_scheduler=True-torch_compile=False]"
```

## Summary by CodeRabbit

* **New Features**
  * Optional Top‑K routing inputs (ids + weights) and dual Top‑K output formats; routing no longer requires per‑token logits.

* **Improvements**
  * Kernelized histogram init and unified Top‑K handling across FP4/FP8/BF16 with device-aware allocations and preserved legacy score path.
  * Routing output dtype now configurable (bf16 or fp32); autotuning/custom ops propagate Top‑K data.

* **Tests**
  * Expanded unit tests to exercise Top‑K‑as‑input flows across backends.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by DomBrown — APPROVED** [review]
Approving the MoE autotune parts as we discussed on Gitlab. Others should review the other bits as you suggested :)

**Review by djns99 — APPROVED** [review]
The routing changes make sense to me. Is this something we should be supporting in the CUTLASS backend, or is this just a quirk of trtllm-gen?

**Review by MatthiasKohl — APPROVED** [review]
LGTM, interface change seems reasonable, and I didn't spot any issue (I didn't go into all details though). Thanks @ChristinaZ

**Review by byshiue — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py [hunk 1]
```diff
@@ -215,25 +218,30 @@ def get_tuning_config(cls) -> TuningConfig:
...
-                               num_experts: int, top_k: int,
-                               n_group: Optional[int],
-                               topk_group: Optional[int],
-                               intermediate_size: int, local_expert_offset: int,
-                               local_num_experts: int,
-                               routed_scaling_factor: Optional[float],
-                               routing_method_type: int,
-                               do_finalize: bool) -> List[torch.Tensor]:
+def fp4_block_scale_moe_runner(
+        routing_logits: Optional[torch.Tensor],
```
Comment (litaotju) [inline tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py:222]: Normally I think the optional inputs should be at the end of inpur args, and have default values. 

Do we have some reasons to put them into the first 2?

File: tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py [hunk 2]
```diff
@@ -245,8 +253,17 @@ def fp4_block_scale_moe_runner(routing_logits: torch.Tensor,
         local_expert_offset, local_num_experts, routed_scaling_factor,
         tile_tokens_dim, routing_method_type, do_finalize)
 
-    inputs = [
-        routing_logits,
+    # Use dummy routing logits for autotuner
+    if routing_logits is None:
+        routing_logits_for_tuner = torch.randn(hidden_states.shape[0],
```
Comment (litaotju) [inline tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py:258]: shall we just the `num_tokens`? Since its already in line 248.

File: tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py [hunk 3]
```diff
@@ -263,36 +280,39 @@ def fp4_block_scale_moe_runner(routing_logits: torch.Tensor,
         "trtllm::fp4_block_scale_moe_runner",
         [kernel_runner],
         kernel_runner.tuning_config,
-        inputs,
+        input_tensors_for_tuner,
     )
 
-    return kernel_runner(inputs, tactic=best_tactic)
+    input_tensors = input_tensors_for_tuner + [topk_weights, topk_ids]
+    input_tensors[
```
Comment (litaotju) [inline tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py:287]: This seems wired. Means we have to call torch.randn even if its not during tuning?

File: cpp/tensorrt_llm/kernels/customMoeRoutingKernels.cu [hunk 4]
```diff
@@ -38,17 +38,18 @@ static constexpr int WARPS_PER_BLOCK = BLOCK_SIZE / WARP_SIZE;
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-template <typename T>
-__device__ T calcSoftmax(cg::thread_block_tile<WARP_SIZE> const& warp, T score, int32_t laneIdx, int32_t NumTopExperts)
+template <typename DataType>
```
Comment (litaotju) [inline cpp/tensorrt_llm/kernels/customMoeRoutingKernels.cu:41]: A nit. Whats wrong with "T"? I think normally it just means data type.

File: tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py (hunk 3 above)
Comment (ChristinaZ) [reply tensorrt_llm/_torch/custom_ops/trtllm_gen_custom_ops.py:287]: Totally agree, let me revise it.

File: cpp/tensorrt_llm/kernels/customMoeRoutingKernels.cu [hunk 5]
```diff
@@ -38,17 +38,18 @@ static constexpr int WARPS_PER_BLOCK = BLOCK_SIZE / WARP_SIZE;
...
+    DataType maxScore = DataType{-INFINITY};
     if (laneIdx < NumTopExperts)
     {
         maxScore = score >= maxScore ? score : maxScore;
     }
-    maxScore = cg::reduce(warp, maxScore, cg::greater<T>());
+    maxScore = cg::reduce(warp, maxScore, cg::greater<DataType>());
 
-    float sumScore{0.f};
+    float sumScore = float{0.f};
```
[the remark this answers is not in the document]
Comment (ChristinaZ) [reply cpp/tensorrt_llm/kernels/customMoeRoutingKernels.cu]: Done.

File: tensorrt_llm/_torch/models/modeling_gpt_oss.py [hunk 6]
```diff
@@ -151,7 +151,10 @@ def __init__(
         )
 
         self.routing_method = RenormalizeMoeRoutingMethod(
-            top_k=pretrained_config.num_experts_per_tok)
+            top_k=pretrained_config.num_experts_per_tok,
+            output_dtype=torch.bfloat16
```
Comment (hlu1) [inline tensorrt_llm/_torch/models/modeling_gpt_oss.py:155]: This looks like a requirement for input dtype by the MoE class. Do you think it's better to set the `output_dtype` in the `TRTLLMGenFusedMoE.__init__` method while keeping the default `torch.float32`? Then this won't spill over into model code and it'll be contained in the MoE class. And you won't need to repeat it in every model file.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py [hunk 7]
```diff
@@ -216,6 +217,66 @@ def forward_impl(
             topk_group = None
             routed_scaling_factor = None
 
+        # Don't support post-quant allgather for fp8 block scale and has_w4a16_mxfp4 for now.
+        is_post_quant_allgather_supported = self.has_nvfp4 or self.has_w4a8_mxfp4_fp8 or self.has_w4a8_mxfp4_mxfp8
+        run_post_quant_allgather = self.use_dp and self.parallel_size > 1 and is_post_quant_allgather_supported
+
+        x_sf = None
```
Comment (hlu1) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py:224]: It's not great to copy-paste these code too many times. We need to think about how to refactor the code better.

If we were going to add more max-throughput optimizations to the trtllm-gen MoE, this class will need to be re-structured. The code is starting to get messy.

**Review by hlu1 — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py [hunk 8]
```diff
@@ -488,17 +488,19 @@ def _(input, residual, norm_weight, expanded_idx_to_permuted_idx,
...
             sz, dtype=torch.int32), router_logits.new_empty(sz,
-                                                            dtype=torch.float32)
+                                                            dtype=output_dtype)
 
     @torch.library.register_fake("trtllm::default_moe_routing_op")
-    def _(router_logits, topk):
+    def _(router_logits, topk, output_dtype: torch.dtype = None):
         num_tokens = router_logits.shape[0]
         sz = (num_tokens, topk)
+        output_dtype = output_dtype if output_dtype is not None else torch.float32
```
[the remark this answers is not in the document]
Comment (ChristinaZ) [reply tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py:503]: Done.

File: tensorrt_llm/_torch/models/modeling_gpt_oss.py (hunk 6 above)
Comment (ChristinaZ) [reply tensorrt_llm/_torch/models/modeling_gpt_oss.py:155]: This part does not look that straightforward.

For the MoE TRTLLM backend, when `run_post_quant_allgather` is set to true, we first run the `RenormalizeMoeRoutingMethod()`. After that, we execute `TRTLLMGenFusedMoE`, which directly uses the output of `RenormalizeMoeRoutingMethod()` and skips the routing part within `TRTLLMGenFusedMoE`. 

When we use the MoE TRTLLM backend, we require the datatype of `topk_values` to be `bfloat16`. While when we use the MoE CUTLASS backend, we require the input data type of `topk_values` to be `float32`. So the output data type of `RenormalizeMoeRoutingMethod()` relies on the workflow. 

So if I understand correctly, I still need to add this part in the model file. If we separate the alltoall/allgather and post_quant logic into a separate function, maybe we can do that.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py (hunk 7 above)
Comment (ChristinaZ) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py:224]: Maybe we can add a separate function including the alltoall/allgather and post_quant logic? So that MoE TRTLLM and CUTLASS backend can share this part. I can discuss with Bo and Anthony about this part.

**Review by litaotju — APPROVED** [review]
